packages/react/src/components/TextArea/TextArea.tsx

```diff
@@ -159,32 +165,122 @@ const TextArea = React.forwardRef((props: TextAreaProps, forwardRef) => {
   const prefix = usePrefix();
   const { isFluid } = useContext(FormContext);
   const { defaultValue, value } = other;
+
+  function getInitialTextCount(
+    value: string | number | undefined,
+    defaultValue: string | number | undefined,
+    counterMode: 'character' | 'word' | undefined
+  ): number {
+    if (defaultValue) {
+      if (counterMode === 'character') {
+        return defaultValue.toString().length;
+      } else {
+        return defaultValue.toString().match(/\w+/g)?.length || 0;
+      }
+    } else if (value) {
+      if (counterMode === 'character') {
+        return value.toString().length;
+      } else {
+        return value.toString().match(/\w+/g)?.length || 0;
+      }
+    }
+    return 0;
+  }
+
   const [textCount, setTextCount] = useState(
-    defaultValue?.toString()?.length || value?.toString()?.length || 0
+    getInitialTextCount(defaultValue, value, counterMode)
   );
   const { current: textAreaInstanceId } = useRef(getInstanceId());
 
   useEffect(() => {
-    setTextCount(
-      defaultValue?.toString()?.length || value?.toString()?.length || 0
-    );
-  }, [value, defaultValue]);
+    setTextCount(getInitialTextCount(defaultValue, value, counterMode));
+  }, [value, defaultValue, counterMode]);
```

Comment:
@francinelucca hmm I would expect this to handle the recounting of characters

Reply:
I think we uncovered a big bug with text area itself, value/defaultValue prop is undefined so when we type the count changes but as soon as you change the mode it tries to calculate initial again and returns 0, that useEffect runs also everytime you type so it gets set to 0 and onChange changes it back to correct value, what was the idea of the useEffect? It was implemented before and I just modified it slightly

Reply:
Ok the issue is that if value or default value is undefined the change of mode reruns the getInitial count and returns 0, I also discovered that everytime we type the initial count function is reexecuted, that is because there is a re-render of the component (for some reason) and useState reruns the initial value, got a fix I think.